src/kernel/bitcoinkernel.h

```diff
@@ -0,0 +1,1145 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_KERNEL_BITCOINKERNEL_H
+#define BITCOIN_KERNEL_BITCOINKERNEL_H
+
+#ifndef __cplusplus
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#else
+#include <cstddef>
+#include <cstdint>
+#endif // __cplusplus
+
+
+#if !defined(BITCOINKERNEL_GNUC_PREREQ)
+#if defined(__GNUC__) && defined(__GNUC_MINOR__)
+#define BITCOINKERNEL_GNUC_PREREQ(_maj, _min) \
+    ((__GNUC__ << 16) + __GNUC_MINOR__ >= ((_maj) << 16) + (_min))
+#else
+#define BITCOINKERNEL_GNUC_PREREQ(_maj, _min) 0
+#endif
+#endif
+
+/* Warning attributes */
+#if defined(__GNUC__) && BITCOINKERNEL_GNUC_PREREQ(3, 4)
+#define BITCOINKERNEL_WARN_UNUSED_RESULT __attribute__((__warn_unused_result__))
+#else
+#define BITCOINKERNEL_WARN_UNUSED_RESULT
+#endif
+#if !defined(BITCOINKERNEL_BUILD) && defined(__GNUC__) && BITCOINKERNEL_GNUC_PREREQ(3, 4)
+#define BITCOINKERNEL_ARG_NONNULL(_x) __attribute__((__nonnull__(_x)))
```

Comment:
When I apply this to the first commit I get:
```
In file included from /home/drgrid/bitcoin/src/test/kernel/test_kernel.cpp:5:
/home/drgrid/bitcoin/src/kernel/bitcoinkernel.h:201:32: error: too many arguments provided to function-like macro invocation
  201 | ) BITCOINKERNEL_ARG_NONNULL(1, 3);
      |                                ^
/home/drgrid/bitcoin/src/kernel/bitcoinkernel.h:34:9: note: macro 'BITCOINKERNEL_ARG_NONNULL' defined here
   34 | #define BITCOINKERNEL_ARG_NONNULL(_x) __attribute__((__nonnull__(_x)))
      |         ^
/home/drgrid/bitcoin/src/kernel/bitcoinkernel.h:201:3: error: expected function body after function declarator
  201 | ) BITCOINKERNEL_ARG_NONNULL(1, 3);
      |   ^
```
Which makes sense, because the macro only expects one argument. I'm not sure how safe it is to make it take a string or a list instead.

Reply:
Sorry, I forgot to include the diff which updates the macro (and 1 instance to show it compiles):

<details>
<summary>git diff on 6c9121f790</summary>

```diff
diff --git a/src/kernel/bitcoinkernel.h b/src/kernel/bitcoinkernel.h
index 9e6bf127db..67248349e2 100644
--- a/src/kernel/bitcoinkernel.h
+++ b/src/kernel/bitcoinkernel.h
@@ -31,9 +31,9 @@
 #define BITCOINKERNEL_WARN_UNUSED_RESULT
 #endif
 #if !defined(BITCOINKERNEL_BUILD) && defined(__GNUC__) && BITCOINKERNEL_GNUC_PREREQ(3, 4)
-#define BITCOINKERNEL_ARG_NONNULL(_x) __attribute__((__nonnull__(_x)))
+#define BITCOINKERNEL_ARG_NONNULL(...) __attribute__((__nonnull__(__VA_ARGS__)))
 #else
-#define BITCOINKERNEL_ARG_NONNULL(_x)
+#define BITCOINKERNEL_ARG_NONNULL(...)
 #endif
 
 #ifdef __cplusplus
@@ -522,7 +522,7 @@ bool BITCOINKERNEL_WARN_UNUSED_RESULT kernel_verify_script(
     unsigned int input_index,
     unsigned int flags,
     kernel_ScriptVerifyStatus* status
-) BITCOINKERNEL_ARG_NONNULL(1) BITCOINKERNEL_ARG_NONNULL(3);
+) BITCOINKERNEL_ARG_NONNULL(1, 3);
 
 /**
  * @brief This disables the global internal logger. No log messages will be

```
</details>

Based on https://gcc.gnu.org/onlinedocs/cpp/Variadic-Macros.html, this should be standard for C99?

Reply:
Thanks, decided to take this. I was a bit careful here, because I lifted the check from secp, which also does not use variadic args: https://github.com/bitcoin-core/secp256k1/blob/master/include/secp256k1.h#L174. But thinking a bit more about it, I could not come up with a good reason not to, so took your suggestion.